File: java/0020. Valid Parentheses.java
package leetcode.java;

import java.util.Stack;

/**
 * Obvious implementation
 * Time complexity: O(N^2)
 * Space complexity: O(N)   
 */
class Solution {
    public boolean isValid(String s) {
        while (s.contains("{}") || s.contains("()") || s.contains("[]")) {
            s = s.replace("{}", "");
            s = s.replace("[]", "");
            s = s.replace("()", "");
        }
        return s.length() == 0;
    }
}


/**
 * Efficient implementation
 * Time complexity: O(N)
 * Space complexity: O(N)   
 */
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack =  new Stack<>();
        for (Character c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.add(c);
            } else {
                if (stack.size() == 0) {
                    return false;
                }
                char temp = stack.pop();
                if (temp == '(' && c != ')') {
                    return false;
                } else if (temp == '{' && c != '}') {
                    return false;
                } else if (temp == '[' && c != ']') {
                    return false;
                }
            }
        }
        return stack.size() == 0;
    }
}

File: python/0020. Valid Parentheses.py
'''
Obvious implementation
Time complexity: O(N^2)
Space complexity: O(N)
'''
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        while "[]" in s or "{}" in s or "()" in s:
            s = s.replace("[]", "")
            s = s.replace("{}", "")
            s = s.replace("()", "")

        return len(s) == 0

'''
Efficient implementation
Time complexity: O(N)
Space complexity: O(N)
'''
class Solution(object):
    def isValid(self, s):
        opening = ['(', '{', '[']

        stack = []

        for bracket in s:
            if bracket in opening:
                stack.append(bracket)
            else:
                if len(stack) == 0:
                    return False
                brack = stack.pop()
                if bracket == ')' and brack != '(':
                    return False
                elif bracket == '}' and brack != '{':
                    return False
                elif bracket == ']' and brack != '[':
                    return False
        
        return len(stack) == 0